File: arxiv.ts
/**
 * This module provides nomralization and validation of the form of the ArXiv identifier.
 * 
 * @module metadatatools
 */

export const ARXIVPattern: string =
  "^arxiv:(\\d{4}\\.\\d{4,5}(v\\d+)?|[a-z\\-]+\\/\\d{7}(v\\d+)?)$";
export const reARXIV: RegExp = new RegExp(ARXIVPattern, "i");
export const newARXIVPattern: string = "^arxiv:\\d{4}\\.\\d{4,5}(v\\d+)?$";
export const reNewARXIV: RegExp = new RegExp(newARXIVPattern, "i");
export const oldARXIVPattern: string = "^arxiv:[a-z\\-]+\\/\\d{7}(v\\d+)?$";
export const reOldARXIV: RegExp = new RegExp(oldARXIVPattern, "i");

/**
 * normalizes the arxiv identifier string
 * @param arxivID 
 * @returns stirng with the normalized ArXiv identifier
 * 
 * @example
 * ```ts
 *   const arxiv: string = normalizeArXivID('ARXIV:2412.03631');
 * 
 *   console.log(arxiv); // displays arXiv:2412.03631
 * ```
 */
export function normalizeArXivID(arxivID: string): string {
  return arxivID.trim().toLowerCase();
}

/**
 * noarmlizes the arxiv identifier
 * @param arxivID 
 * @returns stirng with the normalized ArXiv identifier
 * 
 * @example
 * ```ts
 *   const arxiv: string = normalizeArXivID('ARXIV:2412.03631');
 * 
 *   if (validateArXivID(arxiv)) {
 *       console.log(`${arxiv} appears valid`);
 *   } else {
 *       console.error(`${arxiv}, failed validation`);
 *   }
 * ```
 */
export function validateArXivID(arxivID: string): boolean {
  const normalizedID = normalizeArXivID(arxivID);
  return reARXIV.test(normalizedID);
}

File: arxiv.go
package metadatatools

import (
	"regexp"
	"strings"
)

// ARXIVPattern matches a normalized arXiv identifier in either the new
// (YYMM.NNNNN) or old (archive/YYMMNNN) form, with an optional version
// suffix.
const ARXIVPattern = `^arxiv:(\d{4}\.\d{4,5}(v\d+)?|[a-z\-]+/\d{7}(v\d+)?)$`

// reARXIV is the compiled, case-insensitive form of ARXIVPattern.
var reARXIV = regexp.MustCompile(`(?i)` + ARXIVPattern)

// NewARXIVPattern matches a normalized arXiv identifier in the new
// (post March 2007) YYMM.NNNNN form.
const NewARXIVPattern = `^arxiv:\d{4}\.\d{4,5}(v\d+)?$`

// reNewARXIV is the compiled, case-insensitive form of NewARXIVPattern.
var reNewARXIV = regexp.MustCompile(`(?i)` + NewARXIVPattern)

// OldARXIVPattern matches a normalized arXiv identifier in the old
// (pre March 2007) archive/YYMMNNN form.
const OldARXIVPattern = `^arxiv:[a-z\-]+/\d{7}(v\d+)?$`

// reOldARXIV is the compiled, case-insensitive form of OldARXIVPattern.
var reOldARXIV = regexp.MustCompile(`(?i)` + OldARXIVPattern)

// NormalizeArXivID trims whitespace from arxivID and lower cases it.
//
// Example:
//
//	id := NormalizeArXivID("ARXIV:2412.03631")
//	// id == "arxiv:2412.03631"
func NormalizeArXivID(arxivID string) string {
	return strings.ToLower(strings.TrimSpace(arxivID))
}

// ValidateArXivID normalizes arxivID then validates it against ARXIVPattern.
func ValidateArXivID(arxivID string) bool {
	return reARXIV.MatchString(NormalizeArXivID(arxivID))
}
